src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -319,6 +319,15 @@ private function process_directives_args( string $html, array &$context_stack, a
 					array( $p, &$context_stack, &$namespace_stack, &$tag_stack )
 				);
 			}
+
+			// Remove context from stack if it is a void tag.
+			if ( WP_HTML_Processor::is_void( $tag_name ) ) {
+				foreach ( $directives_prefixes as $directive_prefix ) {
+					if ( 'data-wp-context' === $directive_prefix ) {
+						array_pop( $context_stack );
+					}
```

Comment:
I believe that this is a surprise way to handle this based on how the directive handlers work. that is, they all handle their own cleanup, except now the specific context directive is coupled into the general directive dispatch.

cc: @ockham 

what I think could work more generally is to do two things:
 - when matched on void or special elements (elements without a closing tag) we should run the handlers in reverse as is done above in line 307.
 - update the handlers so that instead of checking `is_closing_tag()` and using HTML semantics, they check something different, such as `'enter'` or `'exit'`, representing whether the code is entering or existing a directive's scope

this implies of course that this function would then pass in that action to the directive handlers.

```php
if ( ! $this->has_and_visits_its_closing_tag() ) {
  call_user_func_array(
	  $func,
	  array( $p, 'enter', &$context_stack, &$namespace_stack, &$tag_stack )
  );
  call_user_func_array(
	  $func,
	  array( $p, 'exit', &$context_stack, &$namespace_stack, &$tag_stack )
  );
} else {
  call_user_func_array(
	  $func,
	  array( $p, $p->is_tag_closer() ? 'exit' : 'enter', &$context_stack, &$namespace_stack, &$tag_stack )
  );
}
```

Reply:
I believe it makes sense to use a more general mechanism and not something hardcoded for the context.

I implemented a different solution based on your feedback in [this commit](https://github.com/WordPress/wordpress-develop/pull/6267/commits/4122ba44b824ec65f623f7395809d28af056a0ec). For void tags or special elements, it runs the directives in both orders.

It seems to be working fine, but I'd love to know if that's what you had in mind.

By the way, I am not convinced about the variable name `$is_exiting_tag`. I can easily changed that if we pursue this approach.

Reply:
I agree this is a better solution, but I wouldn't mind to include this fix in 6.5 as it is, and work on this improvement for a future version because everything in this logic is still private and subject to change 🙂